File: gclib/src/main/java/com/github/guqt178/helper/RouterFragment.java
package com.github.guqt178.helper;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.util.SparseArray;

import java.util.Random;

/**
 * 无界面fragment
 * 优雅地申请权限和处理 onActivityResult
 * 不需要重写Activity的onActivityResult
 *
 */
public class RouterFragment extends Fragment {

    private SparseArray<OnResult> mCallbacks = new SparseArray<>();
    private Random mCodeGenerator = new Random();

    public RouterFragment() {}


    public static RouterFragment newInstance() {
        return new RouterFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    public void startActivityForResult(Intent intent, OnResult callback) {
        int requestCode = makeRequestCode();
        mCallbacks.put(requestCode, callback);
        startActivityForResult(intent, requestCode);
    }

    private int makeRequestCode() {
        int requestCode;
        int tryCount = 0;
        do {
            requestCode = mCodeGenerator.nextInt(0x0000FFFF);
            tryCount++;
        } while (mCallbacks.indexOfKey(requestCode) >= 0 && tryCount < 10);
        return requestCode;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        OnResult callback = mCallbacks.get(requestCode);
        mCallbacks.remove(requestCode);
        if (callback != null) {
            callback.onActivityResult(requestCode,resultCode, data);
        }
    }
}

File: libs/gclib/src/main/kotlin/com/github/guqt178/utils/result/RouterFragment.kt
package com.github.guqt178.utils.result

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.util.SparseArray
import java.util.*

/**
 * 无界面fragment
 * 优雅地申请权限和处理 onActivityResult
 * 不需要重写Activity的onActivityResult
 *
 */
class RouterFragment : Fragment() {
    private val mCallbacks = SparseArray<OnResult>()
    private val mCodeGenerator = Random()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    fun startActivityForResult(intent: Intent?, callback: OnResult?) {
        val requestCode = makeRequestCode()
        mCallbacks.put(requestCode, callback)
        startActivityForResult(intent, requestCode)
    }

    private fun makeRequestCode(): Int {
        var requestCode: Int
        var tryCount = 0
        do {
            requestCode = mCodeGenerator.nextInt(0x0000FFFF)
            tryCount++
        } while (mCallbacks.indexOfKey(requestCode) >= 0 && tryCount < 10)
        return requestCode
    }

    override fun onActivityResult(
        requestCode: Int,
        resultCode: Int,
        data: Intent
    ) {
        super.onActivityResult(requestCode, resultCode, data)
        val callback = mCallbacks[requestCode]
        mCallbacks.remove(requestCode)
        callback?.onActivityResult(requestCode, resultCode, data)
    }

    companion object {
        fun newInstance(): RouterFragment {
            return RouterFragment()
        }
    }
}